junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/TestFactoryTestDescriptor.java

```diff
@@ -87,14 +87,17 @@ protected void invokeTestMethod(JupiterEngineExecutionContext context, DynamicTe
 				context.getExtensionRegistry());
 			TestSource defaultTestSource = getSource().orElseThrow(
 				() -> new JUnitException("Illegal state: TestSource must be present"));
+			// since 5.4: handle single node case (#1521)
+			if (testFactoryMethodResult instanceof DynamicNode) {
```

Comment:
This could be handled in `toDynamicNodeStream` by returning `Stream.of((DynamicNode) testFactoryMethodResult)`.

Reply:
Why wrap the result in a stream only to unwrap two lines later?

Reply:
Moved logic to `toDynamicNodeStream`.